pc2ccs/pc2v9
# I267 shadow GUI update delays

### Description of what the PR does
  Provides a workaround for the issue of Shadow GUIs with huge (10+ minute) delays.   The workaround is to comment-out the statement in the `ServicesView.setContestAndController()` method which constructs a LogWindow.  See **Additional Context**, below, for a more detailed description of why this works.

In addition to commenting-out the line which constructs the LogWindow, the PR also does two other things:  

- reduces the value of the constant `REMOTE_EVENT_FEED_DELAYMS` in `RemoteEventFeedMonitor`, from 500 (msec) to 100 (msec).  There is no longer a need to wait so long between subsequent events in the `RemoteEventFeedMonitor`.   (It could be argued that the value could be reduced even lower, but the value I tested with was 100 so I know that works.)
- removes a reference in `samps/uiname.properties` to an obsolete (now deleted) class.  Note that this change shows as an _addition_ of lines to `uiname.properties`; this is because the machine from which I did the original PR push did not have _any_ value for the `feeder` ui class name (because it was out of date).

### Issue which the PR fixes
Fixes #267 .

### Environment in which the PR was developed (OS,IDE, Java version, etc.)
Windows 10.1, Java 1.8.271

### Precise steps for _testing_ the PR (i.e., how to demonstrate that it works correctly)

- Start a server loaded with a contest for which there exists a remote CCS with the same contest configuration (e.g. https://github.com/icpcsysops/ccsconfig-dhaka/tree/master/contests/pretest2/config).
- Start an Admin, and use it to:
  - Set Start Time to "Undefined
  - If using the above contest, set an unused team (e.g. team 649) to have an alias of "domjudge".
  - Start the contest
- Start an EF Client
  - On the EF Client, select "Shadow Mode" and then "Start Shadowing".
- Click the "Compare Runs" button to display the Runs grid.
- Click the "Compare Scoreboards" button to display the Scoreboard Comparison grid.
- Use the scroll bars to scroll up/down in the grids.
- Click "Refresh" on each grid, several times.

Previously, the Runs Grid and the Scoreboard Grid would not even display for several minutes (at best).  Now, they should display immediately, moving the scroll bars should produce responsive scrolling, and clicking "Refresh" should refresh the screen after a (very short) delay.

### Additional Context

While working on this Issue/PR, I started with the premise that something was bogging down the AWT thread;  running JVisualVM against the Shadow Client (which is actually just a tab on the "ServicesView") demonstrated this was the case -- the moment I clicked "Start Shadowing", the AWT thread went busy and stayed busy for a LONG time (like, 10-15 minutes).

I wanted to know what exactly the AWT thread was "busy" with.  After some googling, I found a way to attach a listener to the AWT thread that would get called on every AWT Event Dispatch.   (If you care, the code that did that is shown below.)

``` java
    private void addAWTEventListener() {
        long eventMask = 0xffffffff ;
        Toolkit.getDefaultToolkit().addAWTEventListener(new MyListener(), eventMask);
    }
    private static class MyListener implements AWTEventListener {
        public void eventDispatched(AWTEvent event) {
            System.out.printf("Got AWT Event %x with params %s from source %s \n", event.getID(), event.paramString(), event.getSource());
        }
    }
```
This code dumps every single AWT event onto the console.  I put a call to method `addAWTEventListener()` in the constructor for the ShadowControlPane.

When I ran with this code, I still got the previous GUI delays -- but I also got every AWT event displayed on the console.  Examining those, I noticed a HUGE number (thousands+) of messages like:

```
Got AWT Event 65 with params COMPONENT_RESIZED (0,0 3035x80000) from source com.ibm.webrunner.j2mclb.MultiColumnListboxBase$ViewArea[,0,0,3035x80000,invalid,alignmentX=0.0,alignmentY=0.0,border=,flags=0,maximumSize=,minimumSize=,preferredSize=]
Got AWT Event 65 with params COMPONENT_RESIZED (0,0 3035x20) from source com.ibm.webrunner.j2mclb.CaptionBar[count=5,captions={Date/Time,Level,Thread,Method,Message}]
Got AWT Event 65 with params COMPONENT_RESIZED (0,0 2626x20) from source com.ibm.webrunner.j2mclb.ListboxCaption[label=Message,length=2626,resizable,showing]
Got AWT Event 65 with params COMPONENT_RESIZED (0,0 3035x80000) from source com.ibm.webrunner.j2mclb.MultiColumnListboxBase$ViewArea[,0,0,3035x80000,invalid,alignmentX=0.0,alignmentY=0.0,border=,flags=0,maximumSize=,minimumSize=,preferredSize=]
Got AWT Event 65 with params COMPONENT_RESIZED (0,0 3035x20) from source com.ibm.webrunner.j2mclb.CaptionBar[count=5,captions={Date/Time,Level,Thread,Method,Message}]
Got AWT Event 65 with params COMPONENT_RESIZED (0,0 2626x20) from source com.ibm.webrunner.j2mclb.ListboxCaption[label=Message,length=2626,resizable,showing] 
```
Notice that these messages refer to resizing different components in a `com.ibm.webrunner.j2mclb` -- in other words, components associated with our "MCLB" class.   

I was stumped for a couple of days, because the Shadow does not use any MCLB instances.  However, after many hours tracing through code, I discovered the following:  
 - the `ServicesView` (which is the frame that contains the tabbed pane containing the ShadowControlPane) has (like most of our Views) a `setContestAndController()` method.  
 - When that method is called (somewhere during initialization of the ServicesView frame), it calls `SwingUtilities.invokeLater()`, passing to it a `run()` method which among other things contains a call to `controller.startLogWindow()`.   
 - Method `startLogWindow()` in turn calls `loadUIClass("LogWindow")`, which in turn calls `LogWindow.newInstance()`.

The definition of the `newInstance()` method includes the following statement:
```
Creates a new instance of the class represented by this Class object. The class is instantiated as if by a new expression with an empty argument list. The class is initialized if it has not already been initialized. 
```
And guess what initializing the `LogWindow` class includes?   It constructs an MCLB to provide a GUI display which tracks log messages!

This LogWindow/MCLB GUI is not visible, because nobody has gone to the Options pane and made it visible (it's not-visible by default) -- but IT EXISTS and IT GETS UPDATED WITH EVERY SINGLE LOG MESSAGE -- which in turn generates a collection of AWT events to refresh/resize/update the various components in the MCLB.

I commented out the call to  `controller.startLogWindow(contest)` in the `ServicesView.setContestAndController()` method, then reran the Shadow -- and BOOM!   I get instantly-responsive GUIs!  (And that's even WITH all the AWT events being dumped to the console -- but the thousands of MCLB updates are no longer there, and the AWT thread can keep up.)

For Dhaka (and hence for BACP in two weeks) a "workaround" is to just do what I did -- comment out the creation of the LogWindow.  It means we wouldn't have the ability to display a GUI view of the log when using the EF Client -- but that seems MUCH less important than having the Shadow being responsive.  Of course we also need to figure out the "proper" fix for this in the long term.
